File: php/tests/StreamParserTest.php
<?php

declare(strict_types=1);

namespace LinkFoundation\LinksNotation\Tests;

use InvalidArgumentException;
use LinkFoundation\LinksNotation\Link;
use LinkFoundation\LinksNotation\Parser;
use LinkFoundation\LinksNotation\StreamParser;
use LogicException;
use PHPUnit\Framework\TestCase;

class StreamParserTest extends TestCase
{
    private const DOCUMENT = "first loves data\n"
        . "profile:\n"
        . "  name Ada\n"
        . "  note \"line one\nline two\"\n"
        . "(nested:\n  child value)\n"
        . 'last sees first';

    /** @param Link[] $links */
    private function render(array $links): array
    {
        return array_map(static fn (Link $link): string => (string) $link, $links);
    }

    public function testMatchesCanonicalParserOneByteAtATime(): void
    {
        $stream = new StreamParser();
        for ($index = 0; $index < strlen(self::DOCUMENT); $index++) {
            $stream->write(self::DOCUMENT[$index]);
        }

        $this->assertSame($this->render((new Parser())->parse(self::DOCUMENT)), $this->render($stream->finish()));
    }

    public function testEmitsOnlyCompleteRecordsToCallback(): void
    {
        $seen = [];
        $stream = (new StreamParser())->onLink(static function (Link $link) use (&$seen): void {
            $seen[] = $link;
        });

        $this->assertSame([], $stream->write("profile:\n  name Ada\n  note \"first\nsecond\"\n"));
        $emitted = $stream->write('n');

        $this->assertCount(1, $emitted);
        $this->assertSame($emitted, $seen);
    }

    public function testSupportsLineChunksFinalRecordAndPosition(): void
    {
        $stream = new StreamParser();
        foreach (preg_split('/(?<=\n)/', self::DOCUMENT, -1, PREG_SPLIT_NO_EMPTY) as $line) {
            $stream->write($line);
        }

        $this->assertSame($this->render((new Parser())->parse(self::DOCUMENT)), $this->render($stream->finish()));
        $this->assertSame(strlen(self::DOCUMENT), $stream->position()['offset']);
        $this->assertSame(0, $stream->position()['buffered']);
    }

    public function testSupportsDrainResetAndBoundedMemory(): void
    {
        $seen = [];
        $stream = (new StreamParser(collect: false, maxBufferSize: 8))->onLink(
            static function (Link $link) use (&$seen): void {
                $seen[] = $link;
            }
        );
        for ($index = 0; $index < 100; $index++) {
            $stream->write("a\n");
        }
        $this->assertSame([], $stream->drain());
        $this->assertCount(1, $stream->finish());
        $this->assertCount(100, $seen);

        $stream->reset();
        $this->expectException(InvalidArgumentException::class);
        $this->expectExceptionMessage('Buffered record');
        $stream->write('123456789');
    }

    public function testProvidesAGeneratorAdapter(): void
    {
        $actual = iterator_to_array(StreamParser::parseChunks(["one link\n", 'two link']));

        $this->assertSame(['(one link)', '(two link)'], $this->render($actual));
    }

    public function testRejectsWritesAfterFinish(): void
    {
        $stream = new StreamParser();
        $stream->finish('one');
        $this->expectException(LogicException::class);
        $stream->write('two');
    }
}

File: java/src/test/java/io/github/linkfoundation/linksnotation/StreamParserTest.java
package io.github.linkfoundation.linksnotation;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import org.junit.jupiter.api.Test;

class StreamParserTest {
  private static final String DOCUMENT =
      """
      first loves data
      # streamed comment
      profile:
        name Ada
        note "line one
      line two"
      (nested:
        child value)
      last sees first""";

  private static List<String> render(List<Link> links) {
    return links.stream().map(Link::toString).toList();
  }

  @Test
  void matchesCanonicalParserOneSymbolAtATime() throws Exception {
    StreamParser stream = new StreamParser();
    DOCUMENT.chars().forEachOrdered(character -> write(stream, Character.toString(character)));

    assertEquals(render(new Parser().parse(DOCUMENT)), render(stream.finish()));
  }

  @Test
  void emitsOnlyCompleteRecords() throws Exception {
    List<Link> seen = new ArrayList<>();
    StreamParser stream = new StreamParser().onLink(seen::add);

    assertEquals(List.of(), stream.write("profile:\n  name Ada\n  note \"first\nsecond\"\n"));
    List<Link> emitted = stream.write("n");

    assertEquals(1, emitted.size());
    assertEquals(emitted, seen);
  }

  @Test
  void supportsLineChunksFinalRecordAndPosition() throws Exception {
    StreamParser stream = new StreamParser();
    for (String line : DOCUMENT.split("(?<=\\n)")) {
      stream.write(line);
    }

    assertEquals(render(new Parser().parse(DOCUMENT)), render(stream.finish()));
    assertEquals(DOCUMENT.length(), stream.position().offset());
    assertEquals(0, stream.position().buffered());
  }

  @Test
  void supportsDrainResetAndBoundedMemory() throws Exception {
    List<Link> seen = new ArrayList<>();
    StreamParser stream = new StreamParser().collect(false).maxBufferSize(8).onLink(seen::add);
    for (int index = 0; index < 100; index++) {
      stream.write("a\n");
    }
    assertEquals(List.of(), stream.drain());
    assertEquals(1, stream.finish().size());
    assertEquals(100, seen.size());

    stream.reset();
    assertThrows(IllegalArgumentException.class, () -> stream.write("123456789"));
  }

  @Test
  void providesLazyAdapters() {
    List<String> links =
        StreamParser.parseChunks(Arrays.asList("one link\n", "two link"))
            .map(Link::toString)
            .toList();

    assertEquals(List.of("(one link)", "(two link)"), links);
  }

  @Test
  void rejectsWritesAfterFinish() throws Exception {
    StreamParser stream = new StreamParser();
    stream.finish("one");

    assertThrows(IllegalStateException.class, () -> stream.write("two"));
  }

  private static void write(StreamParser parser, String chunk) {
    try {
      parser.write(chunk);
    } catch (ParseException exception) {
      throw new AssertionError(exception);
    }
  }
}
